content/en/blogs/understand-requests-and-limits-in-kubernetes.md

```diff
@@ -0,0 +1,310 @@
+description: Understand how Kubernetes requests and limits work and why they can work in the expected way.
+tag: 'Kubernetes, KubeSphere, Requests, Limits'
+createTime: '2021-04-14'
+author: 'Yunkun, Sherlock'
+snapshot: '/images/blogs/en/understand-requests-and-limits-in-kubesphere/cover.png'
+---
+
+As you create resources in a Kubernetes cluster, you may have encountered the following scenarios:
+
+1. No CPU requests or low CPU requests specified for workloads, which means more Pods "seem" to be able to work on the same node. During traffic bursts, your CPU is maxed out with a longer delay while some of your machines may have a CPU soft lockup.
+2. Likewise, no memory requests or low memory requests specified for workloads. Some Pods, especially those running Java business apps, will keep restarting while they can actually run normally in local tests.
+3. In a Kubernetes cluster, workloads are usually not scheduled evenly across nodes. In most cases, in particular, memory resources are unevenly distributed, which means some nodes can see much higher memory utilization than other nodes. As the de facto standard in container orchestration, Kubernetes should have an effective scheduler that ensures the even distribution of resources. But, is it really the case?
+
+Generally, cluster administrators can do nothing but restart the cluster if the above issues happen amid traffic bursts when all of your machines hang and SSH login fails. In this article, we will dive deep into Kubernetes requests and limits by analyzing possible issues and discussing the best practices for them. If you are also interested in the underlying mechanism, you can also find the analysis from the perspective of source code. Hopefully, this article will be helpful for you to understand how Kubernetes requests and limits work, and why they can work in the expected way.
+
+## Concepts
+
+To make full use of resources in a Kubernetes cluster and improve scheduling efficiency, Kubernetes uses requests and limits to control resource allocation for containers. Each container can have its own requests and limits. These two parameters are specified by `resources.requests` and `resources.limits`. Generally speaking, requests are more important in scheduling while limits are more important in running.
+
+```yaml
+resources:  
+    requests:    
+        cpu: 50m
+        memory: 50Mi
+    limits:    
+        cpu: 100m
+        memory: 100Mi
+```
+
+Requests are the smallest amount of resources a container needs. For example, for a container running Spring Boot business, the requests specified must be the minimum amount of resources that a Java Virtual Machine (JVM) needs to consume in the container image. If you set `requests.memory` to `10Mi` for the Pod, obviously, it is not reasonable in this case. The memory resources the JVM is actually using will exceed the value of `requests.memory` (also known as OOM), resulting in the continuous restart of the Pod.
```

Comment:
Precisely speaking, for the above JVM example. If only specified a small value for JVM. There's a high possibility that k8s scheduler will tend to schedule the pod on the node which hasn't sufficient memory to run the JVM. So the pod can't request more memory which JVM bootup process needed. That's the whole story.